src/bin/pg_autoctl/pgsql.c

```diff
@@ -654,43 +721,277 @@ pgsql_drop_replication_slots(PGSQL *pgsql)
+			log_error("Failed to prepare the SQL query for "
+					  "pgsql_replication_slot_maintain");
+			return false;
+		}
+	}
+	return paramIndex;
+}
+
+
+/*
+ * pgsql_replication_slot_drop_removed drop replication slots that belong to
+ * nodes that have been removed. We call that function on the primary, where
+ * the slots are maintained by the replication protocol.
+ *
+ * On the standby nodes, we advance the slots ourselves and use the other
+ * function pgsql_replication_slot_maintain which is complete (create, drop,
+ * advance).
  */
 bool
-postgres_sprintf_replicationSlotName(int nodeId, char *slotName, int size)
+pgsql_replication_slot_drop_removed(PGSQL *pgsql, NodeAddressArray *nodeArray)
 {
-	int bytesWritten =
-		sformat(slotName, size, "%s_%d", REPLICATION_SLOT_NAME_DEFAULT, nodeId);
+	int bytes;
+	char sql[2 * BUFSIZE] = { 0 };
+	char values[BUFSIZE] = { 0 };
 
-	return bytesWritten <= size;
+	/* *INDENT-OFF* */
+	char *sqlTemplate =
+		/*
+		 * We could simplify the writing of this query, but we prefer that it
+		 * looks as much as possible like the query used in
+		 * pgsql_replication_slot_maintain() so that we can maintain both
+		 * easily.
+		 */
+		"WITH nodes(slot_name, lsn) as ("
+		" SELECT '" REPLICATION_SLOT_NAME_DEFAULT "_' || id, lsn"
+		"   FROM (%s) as sb(id, lsn) "
+		"), \n"
+		"dropped as ("
+		" SELECT slot_name, pg_drop_replication_slot(slot_name) "
+		"   FROM pg_replication_slots pgrs LEFT JOIN nodes USING(slot_name) "
+		"  WHERE nodes.slot_name IS NULL "
+		") \n"
+		"SELECT 'drop', slot_name, NULL::pg_lsn FROM dropped";
+	/* *INDENT-ON* */
+
+	Oid paramTypes[NODE_ARRAY_MAX_COUNT * 2] = { 0 };
+	const char *paramValues[NODE_ARRAY_MAX_COUNT * 2] = { 0 };
+	ReplicationSlotMaintainContext context = { 0 };
+
+	int paramCount = BuildNodesArrayValues(nodeArray,
+										   paramTypes, (char **) paramValues,
+										   values, BUFSIZE);
+
+	/* add the computed ($1,$2), ... string to the query "template" */
+	bytes = sformat(sql, 2 * BUFSIZE, sqlTemplate, values);
+
+	if (bytes > 2 * BUFSIZE)
+	{
+		/* shouldn't happen because we only support up to 12 nodes */
+		log_error("Failed to prepare the SQL query for "
+				  "pgsql_replication_slot_maintain");
+		return false;
+	}
+
+	return pgsql_execute_with_params(pgsql, sql,
+									 paramCount, paramTypes, paramValues,
+									 &context,
+									 parseReplicationSlotMaintain);
 }
 
 
 /*
- * pgsql_set_synchronous_standby_names set synchronous_standby_names ton the
- * local Postgres to the value computed on the pg_auto_failover monitor.
+ * pgsql_replication_slot_maintain maintains replication slots that belong to
+ * other standby nodes. We call that function on the standby nodes, where the
+ * slots are maintained manually just in case we need them at failover.
  */
 bool
-pgsql_set_synchronous_standby_names(PGSQL *pgsql,
-									char *synchronous_standby_names)
+pgsql_replication_slot_maintain(PGSQL *pgsql, NodeAddressArray *nodeArray)
```

Comment:
The name of this function and it's comment confused me a bit. I assumed that it only maintained existing slots, but it actually also creates and drops slots. I cannot really think of a better name, but the comment should be a little bit more clear on that, e.g.: 
```
pgsql_replication_slot_maintain creates, updates and drops replication slots that 
belong to other standby nodes...
```

Reply:
Improved the comments, thanks.